Support copy/paste

adlarkin opened the pull request:
Signed-off-by: Ashton Larkin <ashton@openrobotics.org>

# 🎉 New feature

Closes #102

Requires:
- https://github.com/ignitionrobotics/ign-gui/pull/278
- https://github.com/ignitionrobotics/ign-rendering/pull/397
- https://github.com/ignitionrobotics/ign-gazebo/pull/1075

## Summary
This PR builds on top of #959 to provide GUI support for copy/paste. Copy/paste is available via the GUI plugin buttons, right click, and `ctrl-c` + `ctrl-v`.

## Test it
1. Run the following command to start a gazebo world with a model: `ign gazebo pendulum_links.sdf`
1. Load the GUI plugin, and select entities to copy/paste. This can be done while simulation is running or paused:

![copy_paste_demo](https://user-images.githubusercontent.com/42042756/140199534-d52ffdae-1b4b-4c82-b7c8-0013a742fd2c.gif)

## Checklist
- [x] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

chapulina commented on src/gui/plugins/copy_paste/CopyPaste.hh:
```diff
@@ -0,0 +1,59 @@
+{
+  class CopyPastePrivate;
+
+  class CopyPaste : public ignition::gui::Plugin
```
Remember to document the plugin

chapulina commented on src/gui/plugins/copy_paste/CopyPaste.qml:
```diff
@@ -0,0 +1,23 @@
+Rectangle {
+  width: 300
+  height: 200
+  color: "blue"
```
Are you planning to add copy / paste buttons like we have on Gazebo classic?

![image](https://user-images.githubusercontent.com/5751272/132461128-1605f1b7-8f50-4f94-b4ac-650f15e3acae.png)

adlarkin replied:
Yeah, the blue box was just a placeholder while I worked on getting the right click functionality to work. I tried adding the buttons in https://github.com/ignitionrobotics/ign-gazebo/pull/1013/commits/cc3db6b0a6f75d29078a012a156f6e81f4d526f2, but the icons don't appear (the buttons still work - you can see what is copy and what is paste by hovering over the plugin menu). Any idea why the icons aren't visible?

![copy_paste_no_icon_buttons](https://user-images.githubusercontent.com/42042756/132772228-b4730ef3-fb50-4d1d-912c-dee80cda99a1.gif)

chapulina replied:
Do you see any warnings / errors?

adlarkin replied:
The missing icons should be fixed with https://github.com/ignitionrobotics/ign-gazebo/pull/1013/commits/0b5942c37b10115b03ecce760a38b342336a9d90. Should I also follow the [Light Spawning PR](https://github.com/ignitionrobotics/ign-gazebo/pull/587) and add the copy/paste buttons to the [default GUI config](https://github.com/ignitionrobotics/ign-gazebo/blob/main/src/gui/gui.config)?

chapulina commented on src/gui/plugins/scene3d/Scene3D.cc:
```diff
@@ -948,11 +955,22 @@ void IgnRenderer::Render(RenderSync *_renderSync)
       {
         root.Load(this->dataPtr->spawnSdfPath);
       }
+      else if (!this->dataPtr->spawnCloneName.empty())
```
Could you also add this to the new `Spawn` plugin from https://github.com/ignitionrobotics/ign-gazebo/pull/983?

adlarkin commented:
Copy/Paste seems to be working for simple shapes, but I need to figure out a better solution for cloning geometries of cloned visuals (this will be addressed in https://github.com/ignitionrobotics/ign-rendering/pull/397): 
![copy_paste_simple_shapes](https://user-images.githubusercontent.com/42042756/132560335-fab3ab55-dc7f-4b87-b7fc-e0e250a01c0b.gif)

adlarkin commented:
I'm currently facing an issue where an extra object is spawned at the origin after pasting occurs. I have tried to debug this, but cannot figure out what the issue is. The extra entity does not seem to appear in the entity tree, so perhaps there's an issue with cleaning up the preview once pasting occurs? Here's an example of what I am seeing:

![copy_paste_simple_shapes](https://user-images.githubusercontent.com/42042756/132745264-cc9680a2-fa49-4755-838a-e0622aa45db6.gif)

cc @iche033 @chapulina

ahcorde commented on src/gui/plugins/copy_paste/CopyPaste.hh:
```diff
@@ -0,0 +1,81 @@
+    /// \brief Callback for handling a copy service request
+    /// \param[in] _req The service request, which contains the name of the
+    /// entity to be copied
+    /// \param[out] _resp The service response
```
return

ahcorde commented on src/gui/plugins/copy_paste/CopyPaste.hh:
```diff
@@ -0,0 +1,81 @@
+
+    /// \brief Callback for handling a paste service request
+    /// \param[in] _req The service request
+    /// \param[out] _resp The service response
```
return

ahcorde commented on src/gui/plugins/copy_paste/CopyPaste.hh:
```diff
@@ -0,0 +1,81 @@
+
+    /// \internal
+    /// \brief Pointer to private data
+    private: std::unique_ptr<CopyPastePrivate> dataPtr;
```
`IGN_UTILS_IMPL_PTR` ?

adlarkin replied:
I tried using it, but I don't think it will work since some of the things I am using (`transport::node`, `std::mutex`) don't appear to be copyable by default. But, maybe I don't understand `IGN_UTILS_IMPL_PTR` enough. Here's the error output I saw when I tried using `IGN_UTILS_IMPL_PTR`:

<details>
  <summary>Compilation errors</summary>

```
Starting >>> ignition-gazebo6
--- stderr: ignition-gazebo6                             
In file included from /usr/include/ignition/utils1/ignition/utils/ImplPtr.hh:24,
                 from /ws/install/include/ignition/gui6/ignition/gui/GuiEvents.hh:30,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:24:
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh: In instantiation of 'T* ignition::utils::detail::DefaultCopyConstruct(const T&) [with T = ignition::gazebo::CopyPaste::Implementation]':
/usr/include/ignition/utils1/ignition/utils/detail/ImplPtr.hh:120:21:   required from 'ignition::utils::ImplPtr<T> ignition::utils::MakeImpl(Args&& ...) [with T = ignition::gazebo::CopyPaste::Implementation; Args = {}]'
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:58:58:   required from here
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh:56:16: error: use of deleted function 'ignition::gazebo::CopyPaste::Implementation::Implementation(const ignition::gazebo::CopyPaste::Implementation&)'
         return new T(_source);
                ^~~~~~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: note: 'ignition::gazebo::CopyPaste::Implementation::Implementation(const ignition::gazebo::CopyPaste::Implementation&)' is implicitly deleted because the default definition would be ill-formed:
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'ignition::transport::v11::Node::Node(const ignition::transport::v11::Node&)'
In file included from /usr/include/ignition/transport11/ignition/transport.hh:31,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:27:
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: note: 'ignition::transport::v11::Node::Node(const ignition::transport::v11::Node&)' is implicitly deleted because the default definition would be ill-formed:
     class IGNITION_TRANSPORT_VISIBLE Node
                                      ^~~~
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>::unique_ptr(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = ignition::transport::v11::NodePrivate; _Dp = std::default_delete<ignition::transport::v11::NodePrivate>]'
In file included from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/unique_ptr.h:397:7: note: declared here
       unique_ptr(const unique_ptr&) = delete;
       ^~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'std::mutex::mutex(const std::mutex&)'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/mutex:43,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:19:
/usr/include/c++/8/bits/std_mutex.h:97:5: note: declared here
     mutex(const mutex&) = delete;
     ^~~~~
In file included from /usr/include/ignition/utils1/ignition/utils/ImplPtr.hh:24,
                 from /ws/install/include/ignition/gui6/ignition/gui/GuiEvents.hh:30,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:24:
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh: In instantiation of 'void ignition::utils::detail::DefaultCopyAssign(T&, const T&) [with T = ignition::gazebo::CopyPaste::Implementation]':
/usr/include/ignition/utils1/ignition/utils/detail/ImplPtr.hh:120:21:   required from 'ignition::utils::ImplPtr<T> ignition::utils::MakeImpl(Args&& ...) [with T = ignition::gazebo::CopyPaste::Implementation; Args = {}]'
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:58:58:   required from here
/usr/include/ignition/utils1/ignition/utils/detail/DefaultOps.hh:64:15: error: use of deleted function 'ignition::gazebo::CopyPaste::Implementation& ignition::gazebo::CopyPaste::Implementation::operator=(const ignition::gazebo::CopyPaste::Implementation&)'
         _dest = _source;
         ~~~~~~^~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: note: 'ignition::gazebo::CopyPaste::Implementation& ignition::gazebo::CopyPaste::Implementation::operator=(const ignition::gazebo::CopyPaste::Implementation&)' is implicitly deleted because the default definition would be ill-formed:
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'ignition::transport::v11::Node& ignition::transport::v11::Node::operator=(const ignition::transport::v11::Node&)'
In file included from /usr/include/ignition/transport11/ignition/transport.hh:31,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:27:
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: note: 'ignition::transport::v11::Node& ignition::transport::v11::Node::operator=(const ignition::transport::v11::Node&)' is implicitly deleted because the default definition would be ill-formed:
     class IGNITION_TRANSPORT_VISIBLE Node
                                      ^~~~
/usr/include/ignition/transport11/ignition/transport/Node.hh:95:38: error: use of deleted function 'std::unique_ptr<_Tp, _Dp>& std::unique_ptr<_Tp, _Dp>::operator=(const std::unique_ptr<_Tp, _Dp>&) [with _Tp = ignition::transport::v11::NodePrivate; _Dp = std::default_delete<ignition::transport::v11::NodePrivate>]'
In file included from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/unique_ptr.h:398:19: note: declared here
       unique_ptr& operator=(const unique_ptr&) = delete;
                   ^~~~~~~~
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: no matching function for call to 'std::__cxx11::basic_string<char>::operator=(const string&) const'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:664:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(const basic_string& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:664:7: note:   passing 'const string*' {aka 'const std::__cxx11::basic_string<char>*'} as 'this' argument discards qualifiers
/usr/include/c++/8/bits/basic_string.h:732:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(basic_string&& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:732:7: note:   conversion of argument 1 would be ill-formed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: cannot bind rvalue reference of type 'std::__cxx11::basic_string<char>&&' to lvalue of type 'const string' {aka 'const std::__cxx11::basic_string<char>'}
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:703:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(const _CharT* __s)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:714:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(_CharT) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(_CharT __c)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:795:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::initializer_list<_Tp>) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(initializer_list<_CharT> __l)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note: candidate: 'template<class _Tp> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_If_sv<_Tp, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _Tp&) [with _Tp = _Tp; _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
        operator=(const _Tp& __svt)
        ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note:   template argument deduction/substitution failed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: no matching function for call to 'std::__cxx11::basic_string<char>::operator=(const string&) const'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:664:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(const basic_string& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:664:7: note:   passing 'const string*' {aka 'const std::__cxx11::basic_string<char>*'} as 'this' argument discards qualifiers
/usr/include/c++/8/bits/basic_string.h:732:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]' <near match>
       operator=(basic_string&& __str)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:732:7: note:   conversion of argument 1 would be ill-formed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: cannot bind rvalue reference of type 'std::__cxx11::basic_string<char>&&' to lvalue of type 'const string' {aka 'const std::__cxx11::basic_string<char>'}
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/string:52,
                 from /usr/include/c++/8/stdexcept:39,
                 from /usr/include/c++/8/array:39,
                 from /usr/include/c++/8/tuple:39,
                 from /usr/include/c++/8/bits/unique_ptr.h:37,
                 from /usr/include/c++/8/memory:80,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:18:
/usr/include/c++/8/bits/basic_string.h:703:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _CharT*) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(const _CharT* __s)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:714:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(_CharT) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(_CharT __c)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:795:7: note: candidate: 'std::__cxx11::basic_string<_CharT, _Traits, _Alloc>& std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(std::initializer_list<_Tp>) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
       operator=(initializer_list<_CharT> __l)
       ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note: candidate: 'template<class _Tp> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::_If_sv<_Tp, std::__cxx11::basic_string<_CharT, _Traits, _Alloc>&> std::__cxx11::basic_string<_CharT, _Traits, _Alloc>::operator=(const _Tp&) [with _Tp = _Tp; _CharT = char; _Traits = std::char_traits<char>; _Alloc = std::allocator<char>]'
        operator=(const _Tp& __svt)
        ^~~~~~~~
/usr/include/c++/8/bits/basic_string.h:809:8: note:   template argument deduction/substitution failed:
/ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:34:36: error: use of deleted function 'std::mutex& std::mutex::operator=(const std::mutex&)'
 class ignition::gazebo::CopyPaste::Implementation
                                    ^~~~~~~~~~~~~~
In file included from /usr/include/c++/8/mutex:43,
                 from /ws/src/ign-gazebo/src/gui/plugins/copy_paste/CopyPaste.cc:19:
/usr/include/c++/8/bits/std_mutex.h:98:12: note: declared here
     mutex& operator=(const mutex&) = delete;
            ^~~~~~~~
make[2]: *** [src/gui/plugins/copy_paste/CMakeFiles/CopyPaste.dir/CopyPaste.cc.o] Error 1
make[1]: *** [src/gui/plugins/copy_paste/CMakeFiles/CopyPaste.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
make: *** [all] Error 2
---
Failed   <<< ignition-gazebo6 [8.28s, exited with code 2]

Summary: 0 packages finished [8.58s]
  1 package failed: ignition-gazebo6
  1 package had stderr output: ignition-gazebo6
```

</details>

ahcorde commented on src/rendering/SceneManager.cc:
```diff
@@ -393,6 +394,67 @@ rendering::VisualPtr SceneManager::CreateVisual(Entity _id,
+  if (!originalVisual)
+  {
+    ignerr << "Could not find a node with the name [" << _visual
+      << "] in the scene." << std::endl;
```
indent

ahcorde commented on src/rendering/SceneManager.cc:
```diff
@@ -393,6 +394,67 @@ rendering::VisualPtr SceneManager::CreateVisual(Entity _id,
+  if (this->dataPtr->visuals.find(_id) != this->dataPtr->visuals.end())
+  {
+    ignerr << "Entity with Id: [" << _id << "] already exists in the scene"
+           << std::endl;
```
indent

ahcorde commented on src/gui/plugins/scene3d/Scene3D.cc:
```diff
@@ -1153,7 +1171,8 @@ bool IgnRenderer::GeneratePreview(const sdf::Root &_sdf)
   {
-    ignwarn << "Only model entities can be spawned at the moment." << std::endl;
+    ignwarn << "Only model and light entities can be spawned at the moment."
+      << std::endl;
```
indent

ahcorde commented on src/gui/plugins/scene3d/Scene3D.cc:
```diff
@@ -948,11 +955,22 @@ void IgnRenderer::Render(RenderSync *_renderSync)
       {
-        ignwarn << "Failed to spawn: no SDF string or path" << std::endl;
+        ignwarn << "Failed to spawn: no SDF string, path, or name of resource "
+          << "to clone" << std::endl;
```
indent

ahcorde commented on src/gui/plugins/copy_paste/CopyPaste.cc:
```diff
@@ -0,0 +1,165 @@
+        &CopyPaste::CopyServiceCB, this))
+  {
+    ignerr << "Error advertising service [" << this->dataPtr->copyService
+      << "]" << std::endl;
```
indent

ahcorde commented on src/gui/plugins/copy_paste/CopyPaste.cc:
```diff
@@ -0,0 +1,165 @@
+        &CopyPaste::PasteServiceCB, this))
+  {
+    ignerr << "Error advertising service [" << this->dataPtr->pasteService
+      << "]" << std::endl;
```
indent

adlarkin commented:
Another issue I'm trying to troubleshoot is that if I copy/paste an entity, I'll sometimes see the following error message (this is from https://github.com/ignitionrobotics/ign-rendering/pull/397):
```
[GUI] [Err] [BaseGeometry.hh:110] Cloning a geometry failed because the name of the mesh is missing.
```

I usually see this when I try to copy/paste an entity that has already been moved with the transform tool or copy/paste an entity via the plugin buttons. I don't see it happen as much when I copy/paste via the right click menu. I've tried debugging, but I'm not sure why the mesh information doesn't seem to be attached to an entity at times.

iche033 commented:
> I'm currently facing an issue where an extra object is spawned at the origin after pasting occurs. 

this could be the preview visual is not removed properly. I noticed that it's calling `SceneManager::RemoveEntity(_id)` on the model entity Id. That function does not recursively delete all child visuals so you'll likely need to remove other visuals in the model as well. 

> Another issue I'm trying to troubleshoot is that if I copy/paste an entity, I'll sometimes see the following error message (this is from ignitionrobotics/ign-rendering#397):

That turns out to be the WireBox that is attached to the model when it is selected. We'll need to be careful about other visualizations such as joint, CoM, inertia visuals that can be attached to the model visual. So maybe instead of recursively copying all the child visuals, we may need to figure out a way to omit these gui-only visuals. One idea is to tag these gui-only visuals using user-data and in the SceneManager::CopyVisual function, you can selectively clone visuals that do not have this tag.

adlarkin commented:
> you'll likely need to remove other visuals in the model as well.
> 
> That turns out to be the WireBox that is attached to the model when it is selected.

Thanks, @iche033! I have implemented these fixes in https://github.com/ignitionrobotics/ign-gazebo/pull/1013/commits/0b3ab43349805b9921ac33950e162ab5b8ce80ed, and everything seems to work well now with simple shapes and the `Scene3d` plugin (I haven't added copy/paste to the minimal scene yet). This PR now also depends on https://github.com/ignitionrobotics/ign-rendering/pull/442

> We'll need to be careful about other visualizations such as joint, CoM, inertia visuals

I haven't worked on adding ignore clone `UserData` for these yet, but I will work on them next. While I do that, if you'd like to review my most recent changes, feel free to do so.

adlarkin commented:
This is ready for another round of review. I have added some details to the PR description that show a demo and instructions for testing. I have a few questions/notes for my reviewers (@chapulina @iche033):

1. I've added copy/paste options to the right click menu (`EntityContextMenu`). However, these buttons won't work unless the copy/paste plugin has been loaded. Because of this, should I remove the copy/paste options from the right click menu, and only allow users to copy/paste via the plugin buttons themselves?
2. I have added the corresponding server functionality in `Scene3d`, but not to the `Spawn` yet (the `Spawn`/`MinimalScene` changes were made after I started working on this PR). Since there are already quite a few changes here, I will copy the `Scene3d` functionality over to `Spawn` either in a separate PR or once everything else in this PR has been reviewed.

iche033 commented on src/rendering/RenderUtil.cc:
```diff
@@ -2674,6 +2674,7 @@ void RenderUtilPrivate::HighlightNode(const rendering::NodePtr &_node)
     wireBoxVis->SetInheritScale(false);
     wireBoxVis->AddGeometry(wireBox);
     wireBoxVis->SetMaterial(white, false);
+    wireBoxVis->SetUserData("skip-visual-clone", static_cast<bool>(true));
```
do you need to set this to other visuals like joint visuals, CoM, inertia, collision visuals?

adlarkin commented:
I've updated this PR to support copy/paste via `ctrl-c` and `ctrl-v`. The plugin is also automatically loaded with the default gazebo gui config (see the demo in the PR description). This should be ready for final review!
